store/store.go

```diff
@@ -0,0 +1,519 @@
+		s.addEmptyHeight(height)
+		return nil
+	}
+
+	if has, _ := s.hasByHash(datahash); !has {
+		return errors.New("cannot link non-existing file")
+	}
+	return s.createHeightLink(datahash, height)
+}
+
+func (s *Store) createHeightLink(datahash share.DataHash, height uint64) error {
+	// short circuit if link exists
+	if has, _ := s.hasByHeight(height); has {
+		return nil
+	}
+
+	filePath := s.basepath + blocksPath + datahash.String()
+	// create hard link with height as name
+	linkPath := s.basepath + heightsPath + strconv.Itoa(int(height))
+	err := os.Link(filePath, linkPath)
+	if err != nil {
+		return fmt.Errorf("creating hard link: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Store) GetByHeight(ctx context.Context, height uint64) (eds.AccessorStreamer, error) {
+	lock := s.stripLock.byHeight(height)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	f, err := s.getByHeight(height)
+	s.metrics.observeGet(ctx, time.Since(tNow), err != nil)
+	return f, err
+}
+
+func (s *Store) getByHeight(height uint64) (eds.AccessorStreamer, error) {
+	if s.isEmptyHeight(height) {
+		return emptyFile, nil
+	}
+
+	f, err := s.cache.Get(height)
+	if err == nil {
+		return f, nil
+	}
+
+	path := s.basepath + heightsPath + fmt.Sprintf("%d", height)
+	f, err = file.OpenQ1Q4File(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrNotFound
+		}
+		return nil, fmt.Errorf("opening ODS file: %w", err)
+	}
+	return wrappedFile(f), nil
+}
+
+func (s *Store) HasByHash(ctx context.Context, datahash share.DataHash) (bool, error) {
+	if datahash.IsEmptyRoot() {
+		return true, nil
+	}
+	lock := s.stripLock.byDatahash(datahash)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	exist, err := s.hasByHash(datahash)
+	s.metrics.observeHas(ctx, time.Since(tNow), err != nil)
+	return exist, err
+}
+
+func (s *Store) hasByHash(datahash share.DataHash) (bool, error) {
+	if datahash.IsEmptyRoot() {
+		return true, nil
+	}
+	path := s.basepath + blocksPath + datahash.String()
+	return pathExists(path)
+}
+
+func (s *Store) HasByHeight(ctx context.Context, height uint64) (bool, error) {
+	lock := s.stripLock.byHeight(height)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	exist, err := s.hasByHeight(height)
+	s.metrics.observeHas(ctx, time.Since(tNow), err != nil)
+	return exist, err
+}
+
+func (s *Store) hasByHeight(height uint64) (bool, error) {
+	if s.isEmptyHeight(height) {
+		return true, nil
+	}
+
+	_, err := s.cache.Get(height)
+	if err == nil {
+		return true, nil
+	}
+
+	path := s.basepath + heightsPath + fmt.Sprintf("%d", height)
+	return pathExists(path)
+}
+
+func (s *Store) Remove(ctx context.Context, height uint64) error {
+	lock := s.stripLock.byHeight(height)
+	lock.Lock()
+	defer lock.Unlock()
+
+	tNow := time.Now()
+	err := s.remove(ctx, height)
+	s.metrics.observeRemove(ctx, time.Since(tNow), err != nil)
+	return err
+}
+
+func (s *Store) remove(ctx context.Context, height uint64) error {
```

Comment:
The remove logic misses empty height clean up.

Reply:
Honestly, I still feel like the map and a single file approach add complexity to the code for no reason. We need to iterate over all the heights in the file remove the unneeded one in place, and rewrite/reencode the whole thing.


While storing and checking for an empty file is as simple as this:

```go
// createEmptyFile ensures the file exists and is empty
func createEmptyFile(filePath string) error {
	// Create or truncate the file
	file, err := os.Create(filePath)
	if err != nil {
		return err
	}
	// Close the file after creation/truncation
	defer file.Close()
	return nil
}

// checkFileEmpty checks if the given file exists and is empty
func checkFileEmpty(filePath string) (exists bool, empty bool, err error) {
	// Get the file info
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			// File does not exist
			return false, false, nil
		}
		// Other error
		return false, false, err
	}

	// File exists, check if it is empty
	return true, fileInfo.Size() == 0, nil
}
```

Reply:
In the end, the proposed solution will have much less code for empty heights handling, which will never be a bottleneck. Current solution feels overengineered for what we need

Reply:
In real world scenario we would barely have empty heights. The suggested solution would need to perform checkEmptyFile on every file read, even for non empty files.  I tried t avoid by having inmemory map lookup instead, which is much faster and easier, as it does not disk I/O.
Removal is intentionally not implemented. It might looks missing, but there is no actual value to make it. We don't expect empty height and each will take only 32 bytes on disk. Not worth to add complexity to delete this data

Reply:
Sync: decide to go with a single empty file and multiple links